### PR TITLE
Fix Memory Leak in AST Operations

### LINKED ISSUE #3316: Garbage collection: memory leak in SMPC
**Describe the bug**
Garbage collection is not working as expected in SMPC

**To Reproduce**
```python
import torch as th
import syft as sy

hook = sy.TorchHook(th)
bob = sy.VirtualWorker(hook, id="bob")
alice = sy.VirtualWorker(hook, id="alice")
crypto_provider = sy.VirtualWorker(hook, id="james")

torch = th
syft = sy

class Classifier(torch.nn.Module):
    def __init__(self, in_features, out_features):
        super(Classifier, self).__init__()
        self.fc = torch.nn.Linear(in_features, out_features)
    def forward(self, x):
        logits = self.fc(x)
        return logits
    
classifier = Classifier(in_features = 5, out_features = 2)

bob.clear_objects()
classifier = classifier.fix_prec().share(bob, alice, crypto_provider = crypto_provider)
for i in range(3):
    a = torch.ones(1,5)
    b = a.fix_prec().share(bob, alice, crypto_provider = crypto_provider)
    classifier(b)
    print(len(bob._objects))
```

**Expected behavior**
the number of object in bob's store shouldn't increase, but currently it does (quite a lot)

### PR DESCRIPTION
**Pull request by Syzygianinfern0**
Fixes #3316 

## Current progress
- [x] Worker 1
- [x] Worker 2
- [x] Crypto Provider
- [x] Multi Party
- [x] Clean-Up
- [x] Tests

## The Change in Behaviour

<details>
  <summary>Code Used to Test</summary>
  
```python
import torch as th
import syft as sy

hook = sy.TorchHook(th)
alice = sy.VirtualWorker(hook, id="alice")
bob = sy.VirtualWorker(hook, id="bob")
crypto_provider = sy.VirtualWorker(hook, id="james")

torch = th
syft = sy


class Classifier(torch.nn.Module):
    def __init__(self, in_features, out_features):
        super(Classifier, self).__init__()
        self.fc = torch.nn.Linear(in_features, out_features)

    def forward(self, x):
        logits = self.fc(x)
        return logits


def mul(a, b):
    c = a * b
    return c


def add(a, b):
    c = a + b
    return c


alice.clear_objects()
bob.clear_objects()

a = torch.ones(1, 5)
b = torch.ones(1, 5)

a = a.encrypt(workers=[alice, bob], crypto_provider=crypto_provider)
b = b.encrypt(workers=[alice, bob], crypto_provider=crypto_provider)


classifier = Classifier(in_features=5, out_features=2)
classifier = classifier.fix_prec().share(bob, alice, crypto_provider=crypto_provider)


for i in range(3):
    print("-" * 20)
    print(f"Alice: {len(alice._objects)}")
    print(f"Bob: {len(bob._objects)}")
    print(f"Crypto Provider: {len(crypto_provider._objects)}")

    c = classifier(b)  # <<<<< Just Toggle Between Operations
    # c = mul(a, b)
    # c = add(a, b)

    print(f"Alice: {len(alice._objects)}")
    print(f"Bob: {len(bob._objects)}")
    print(f"crypto_provider: {len(crypto_provider._objects)}")

```
</details>

Test

<details>
<summary>Same Table in MD</summary>

#### Sorry for the Ugly Table 😅
(\<br>'s don't go well in GitHub)
<table>
    <tr>
        <th>Operation</th>
        <th>Current Behaviour</th>
        <th>After Merge</th>
    </tr>
    <tr>
        <td>Addition</td>
        <td>
            <code>==========Iter1==========<br>
            Alice: 4<br>
            Bob: 4<br>
            Crypto Provider: 0<br>
            !!OPERATION!!<br>
            Alice: 5<br>
            Bob: 5<br>
            crypto_provider: 0<br>
            ==========Iter2==========<br>
            Alice: 5<br>
            Bob: 5<br>
            Crypto Provider: 0<br>
            !!OPERATION!!<br>
            Alice: 5<br>
            Bob: 5<br>
            crypto_provider: 0<br>
            ==========Iter3==========<br>
            Alice: 5<br>
            Bob: 5<br>
            Crypto Provider: 0<br>
            !!OPERATION!!<br>
            Alice: 5<br>
            Bob: 5<br>
            crypto_provider: 0</code>
        </td>
        <td>
            <code>==========Iter1==========<br>
            Alice: 4<br>
            Bob: 4<br>
            Crypto Provider: 0<br>
            !!OPERATION!!<br>
            Alice: 5<br>
            Bob: 5<br>
            crypto_provider: 0<br>
            ==========Iter2==========<br>
            Alice: 5<br>
            Bob: 5<br>
            Crypto Provider: 0<br>
            !!OPERATION!!<br>
            Alice: 5<br>
            Bob: 5<br>
            crypto_provider: 0<br>
            ==========Iter3==========<br>
            Alice: 5<br>
            Bob: 5<br>
            Crypto Provider: 0<br>
            !!OPERATION!!<br>
            Alice: 5<br>
            Bob: 5<br>
            crypto_provider: 0
            </code>
        </td>
    </tr>
        <tr>
        <td>Multiplication</td>
        <td>
            <code>==========Iter1==========<br>
                Alice: 4<br>
                Bob: 4<br>
                Crypto Provider: 0<br>
                !!OPERATION!!<br>
                Alice: 12<br>
                Bob: 6<br>
                crypto_provider: 1<br>
                ==========Iter2==========<br>
                Alice: 12<br>
                Bob: 6<br>
                Crypto Provider: 1<br>
                !!OPERATION!!<br>
                Alice: 19<br>
                Bob: 7<br>
                crypto_provider: 2<br>
                ==========Iter3==========<br>
                Alice: 19<br>
                Bob: 7<br>
                Crypto Provider: 2<br>
                !!OPERATION!!<br>
                Alice: 26<br>
                Bob: 8<br>
                crypto_provider: 3            
            </code>
        </td>
        <td>
            <code>==========Iter1==========<br>
            Alice: 4<br>
            Bob: 4<br>
            Crypto Provider: 0<br>
            !!OPERATION!!<br>
            Alice: 5<br>
            Bob: 5<br>
            crypto_provider: 0<br>
            ==========Iter2==========<br>
            Alice: 5<br>
            Bob: 5<br>
            Crypto Provider: 0<br>
            !!OPERATION!!<br>
            Alice: 5<br>
            Bob: 5<br>
            crypto_provider: 0<br>
            ==========Iter3==========<br>
            Alice: 5<br>
            Bob: 5<br>
            Crypto Provider: 0<br>
            !!OPERATION!!<br>
            Alice: 5<br>
            Bob: 5<br>
            crypto_provider: 0
            </code>
        </td>
    </tr>
    <tr>
        <td>nn.Linear</td>
        <td>
            <code>==========Iter1==========<br>
            Alice: 4<br>
            Bob: 4<br>
            Crypto Provider: 0<br>
            !!OPERATION!!<br>
            Alice: 9<br>
            Bob: 9<br>
            crypto_provider: 1<br>
            ==========Iter2==========<br>
            Alice: 9<br>
            Bob: 9<br>
            Crypto Provider: 1<br>
            !!OPERATION!!<br>
            Alice: 13<br>
            Bob: 13<br>
            crypto_provider: 2<br>
            ==========Iter3==========<br>
            Alice: 13<br>
            Bob: 13<br>
            Crypto Provider: 2<br>
            !!OPERATION!!<br>
            Alice: 17<br>
            Bob: 17<br>
            crypto_provider: 3
            </code>
        </td>
        <td>
            <code>
            ==========Iter1==========<br>
            Alice: 4<br>
            Bob: 4<br>
            Crypto Provider: 0<br>
            !!OPERATION!!<br>
            Alice: 5<br>
            Bob: 5<br>
            crypto_provider: 0<br>
            ==========Iter2==========<br>
            Alice: 5<br>
            Bob: 5<br>
            Crypto Provider: 0<br>
            !!OPERATION!!<br>
            Alice: 5<br>
            Bob: 5<br>
            crypto_provider: 0<br>
            ==========Iter3==========<br>
            Alice: 5<br>
            Bob: 5<br>
            Crypto Provider: 0<br>
            !!OPERATION!!<br>
            Alice: 5<br>
            Bob: 5<br>
            crypto_provider: 0
            </code>
        </td>
    </tr>
</table>
</details>

### CODE REVIEW

File: syft/frameworks/torch/mpc/beaver.py [hunk 1]
```diff
@@ -39,4 +39,5 @@ def request_triple(
     b_shared = shares[a.numel() : -c.numel()].reshape(b_size)
     c_shared = shares[-c.numel() :].reshape(c.shape)
 
+    shares.get()  # TODO: Workaround to be cleansed
```
Comment (LaRiffle) [inline syft/frameworks/torch/mpc/beaver.py:42]: Is this needed?

File: syft/frameworks/torch/mpc/spdz.py [hunk 2]
```diff
@@ -30,13 +30,13 @@ def spdz_mul(cmd: Callable, x_sh, y_sh, crypto_provider: AbstractWorker, field:
...
+    delta = x_sh - a  # Not gced coz the pointer to other loc doesnt exist
     epsilon = y_sh - b
     # Reconstruct and send to all workers
-    delta = delta.reconstruct()
-    epsilon = epsilon.reconstruct()
+    delta_recon = delta.reconstruct()
+    epsilon_recon = epsilon.reconstruct()
 
-    delta_epsilon = cmd(delta, epsilon)
+    delta_epsilon = cmd(delta_recon, epsilon_recon)
```
Comment (LaRiffle) [inline syft/frameworks/torch/mpc/spdz.py:39]: Here I guess nothing has changed

File: syft/frameworks/torch/mpc/spdz.py [hunk 3]
```diff
@@ -46,7 +46,22 @@ def spdz_mul(cmd: Callable, x_sh, y_sh, crypto_provider: AbstractWorker, field:
...
+    #     loc.de_register_obj(epsilon)
+    # try:
+    #     # delta.locations[0].de_register_obj(delta)
+    #     # delta.locations[0].rm_obj(delta.child[delta.locations[0].id])
+    #     del delta.child[delta.locations[0].id]
+    #     # epsilon.locations[0].de_register_obj(epsilon)
+    #     # epsilon.locations[0].rm_obj(epsilon.child[epsilon.locations[0].id])
+    #     del epsilon.child[epsilon.locations[0].id]
+    # except Exception as e:
+    #     print(e)
```
Comment (LaRiffle) [inline syft/frameworks/torch/mpc/spdz.py:65]: As well here

File: syft/frameworks/torch/tensors/interpreters/additive_shared.py [hunk 4]
```diff
@@ -247,7 +248,7 @@ def reconstruct(self):
         """
         workers = self.locations
 
-        ptr_to_sh = self.wrap().send(workers[0], **no_wrap)
+        ptr_to_sh = self.copy().wrap().send(workers[0], **no_wrap)
```
Comment (LaRiffle) [inline syft/frameworks/torch/tensors/interpreters/additive_shared.py:251]: why do we need copy here?

File: syft/generic/frameworks/hook/hook_args.py [hunk 5]
```diff
@@ -711,6 +711,7 @@ def register_tensor(
         response_ids: List of ids where the tensor should be stored
             and each id is pop out when needed.
     """
+    # owner.de_register_obj(tensor)
```
Comment (LaRiffle) [inline syft/generic/frameworks/hook/hook_args.py:714]: That's not needed finally?

Comment (Syzygianinfern0) [reply syft/generic/frameworks/hook/hook_args.py:714]: Yes. Works without it.
Removed it as it was breaking the inplace operations such as __iadd__, etc

File: syft/frameworks/torch/tensors/interpreters/additive_shared.py (hunk 4 above)
Comment (Syzygianinfern0) [reply syft/frameworks/torch/tensors/interpreters/additive_shared.py:251]: Destroys the original share in the due to ```.remote_get()``` (next line)

File: syft/generic/frameworks/hook/hook_args.py (hunk 5 above)
Comment (karlhigley) [reply syft/generic/frameworks/hook/hook_args.py:720]: If it's not needed, let's remove it.

File: syft/workers/base.py [hunk 6]
```diff
@@ -542,13 +542,12 @@ def execute_communication_action(self, action: CommunicationAction) -> PointerTe
...
             obj = self.get_obj(obj_id)
             response = source_worker.send(obj, *destinations, **kwargs_)
-
-            response = hook_args.register_response("send", response, [sy.ID_PROVIDER.pop()], self)
-
-            # @lariffle: We only remove remote objects when the operations are inplace
-            # otherwise we could have stale pointers which we really want to avoid.
-            # TODO: needs more discussion
-            if kwargs_.get("inplace"):
+            if kwargs_.get("requires_grad", False):
```
Comment (karlhigley) [inline syft/workers/base.py:545]: I don't follow why `requires_grad` controls whether or not we register the response. What's the rationale for that? Can we make it clearer in this code?

File: test/torch/pointers/test_pointer_tensor.py [hunk 7]
```diff
@@ -289,7 +289,7 @@ def test_move(workers):
 
     x.move(alice)
 
-    assert x.id_at_location in bob._objects
+    # assert x.id_at_location in bob._objects
```
Comment (karlhigley) [inline test/torch/pointers/test_pointer_tensor.py:292]: If these assertions are no longer passing with the changes in this PR, can we flip them around to `assert x.id_at_location not in bob._objects`? Seems like that would reflect the intuitive semantics of `move()`.

File: test/torch/tensors/test_additive_shared.py [hunk 8]
```diff
@@ -859,3 +844,37 @@ def test_correct_tag_and_description_after_send(workers):
 
     assert me.request_search("tag_additive_test1", location=alice)
     assert me.request_search("tag_additive_test2", location=alice)
+
+
+def test_garbage_collect_reconstruct(workers):
```
Comment (karlhigley) [inline test/torch/tensors/test_additive_shared.py:849]: ❤️ the new garbage collection tests!

File: test/torch/pointers/test_pointer_tensor.py [hunk 9]
```diff
@@ -554,3 +554,15 @@ def test_setting_back_grad_to_origin_after_move(workers):
         z.backward()
 
         assert (x.grad == th.tensor([4.0, 4.0, 4.0, 4.0, 4.0])).all()
+
+
+def test_iadd(workers):
```
Comment (karlhigley) [inline test/torch/pointers/test_pointer_tensor.py:559]: 👍

File: test/torch/pointers/test_pointer_tensor.py (hunk 7 above)
Comment (Syzygianinfern0) [reply test/torch/pointers/test_pointer_tensor.py:292]: 👍

File: syft/workers/base.py (hunk 6 above)
Comment (Syzygianinfern0) [reply syft/workers/base.py:545]: ### If a tensor requires grad, it mustn't get destroyed till the `loss.backward()` is invoked. 
Basically, we shouldn't delete the tensor when `requires_grad` is `True`, but always when its `None` or `False`. So, for it to be preserved till backprop, the tensor has to be registered.

Comment (karlhigley) [reply syft/workers/base.py:545]: Hmm. Well, maybe we could add a comment that conveys that part?

It would probably be clearer if the flag was called `backprop`, but that change would extend beyond the scope of this PR.

Comment (Syzygianinfern0) [reply syft/workers/base.py:545]: Currently there is another PR in the happening that fixes some GC issues with the backprop and involves this part of the codebase. I will address your comment in them. 👍

File: syft/generic/frameworks/hook/hook_args.py (hunk 5 above)
Comment (Syzygianinfern0) [reply syft/generic/frameworks/hook/hook_args.py]: @karlhigley Done

File: test/torch/pointers/test_pointer_tensor.py (hunk 7 above)
Comment (Syzygianinfern0) [reply test/torch/pointers/test_pointer_tensor.py]: @karlhigley Done

**Comment by AlanAboudib** [comment]
Hooray!! It is merged
